lib/lfrfid/protocols/protocol_h10301.c

```diff
@@ -0,0 +1,386 @@
+    for(int8_t i = 15; i >= 0; i--) {
+        switch((*(card_data + 2) >> (2 * i)) & 0b11) {
+        case 0b01:
+            result = (result << 1) | 0;
+            break;
+        case 0b10:
+            result = (result << 1) | 1;
+            break;
+        default:
+            break;
+        }
+    }
+
+    uint8_t data[H10301_DECODED_DATA_SIZE] = {
+        (uint8_t)(result >> 17), (uint8_t)(result >> 9), (uint8_t)(result >> 1)};
+
+    memcpy(decoded_data, &data, H10301_DECODED_DATA_SIZE);
+}
+
+bool protocol_h10301_decoder_feed(ProtocolH10301* protocol, bool level, uint32_t duration) {
+    bool value;
+    uint32_t count;
+    bool result = false;
+
+    fsk_demod_feed(protocol->decoder.fsk_demod, level, duration, &value, &count);
+    if(count > 0) {
+        for(size_t i = 0; i < count; i++) {
+            protocol_h10301_decoder_store_data(protocol, value);
+            if(protocol_h10301_can_be_decoded(protocol->encoded_data)) {
+                protocol_h10301_decode(protocol->encoded_data, protocol->data);
+                result = true;
+                break;
+            }
+        }
+    }
+
+    return result;
+};
+
+static void protocol_h10301_write_raw_bit(bool bit, uint8_t position, uint32_t* card_data) {
+    if(bit) {
+        card_data[position / H10301_BIT_SIZE] |=
+            1UL << (H10301_BIT_SIZE - (position % H10301_BIT_SIZE) - 1);
+    } else {
+        card_data[position / H10301_BIT_SIZE] &=
+            ~(1UL << (H10301_BIT_SIZE - (position % H10301_BIT_SIZE) - 1));
+    }
+}
+
+static void protocol_h10301_write_bit(bool bit, uint8_t position, uint32_t* card_data) {
+    protocol_h10301_write_raw_bit(bit, position + 0, card_data);
+    protocol_h10301_write_raw_bit(!bit, position + 1, card_data);
+}
+
+void protocol_h10301_encode(const uint8_t* decoded_data, uint8_t* encoded_data) {
+    uint32_t card_data[H10301_DECODED_DATA_SIZE] = {0, 0, 0};
+
+    uint32_t fc_cn = (decoded_data[0] << 16) | (decoded_data[1] << 8) | decoded_data[2];
+
+    // even parity sum calculation (high 12 bits of data)
+    uint8_t even_parity_sum = 0;
+    for(int8_t i = 12; i < 24; i++) {
+        if(((fc_cn >> i) & 1) == 1) {
+            even_parity_sum++;
+        }
+    }
+
+    // odd parity sum calculation (low 12 bits of data)
+    uint8_t odd_parity_sum = 1;
+    for(int8_t i = 0; i < 12; i++) {
+        if(((fc_cn >> i) & 1) == 1) {
+            odd_parity_sum++;
+        }
+    }
+
+    // 0x1D preamble
+    protocol_h10301_write_raw_bit(0, 0, card_data);
+    protocol_h10301_write_raw_bit(0, 1, card_data);
+    protocol_h10301_write_raw_bit(0, 2, card_data);
+    protocol_h10301_write_raw_bit(1, 3, card_data);
+    protocol_h10301_write_raw_bit(1, 4, card_data);
+    protocol_h10301_write_raw_bit(1, 5, card_data);
+    protocol_h10301_write_raw_bit(0, 6, card_data);
+    protocol_h10301_write_raw_bit(1, 7, card_data);
+
+    // company / OEM code 1
+    protocol_h10301_write_bit(0, 8, card_data);
```

Comment:
maybe use a cycle?

Reply:
h10301 needs rework, it will be done later